### PR TITLE
Implement ota_http_proxy server

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Introduction
This PR implements a cache-enabled standalone http proxy server dedicated for streaming OTA updates to subECUs.
This ota_proxy server runs as an independent proxy server. It will cache the OTA files request by URL, and stream the files to the subECUs.

## NOTE
1. https is not supported between mainECU and subECUs as we enable cached.
2. https is supported when mainECU fetch files from the internet.
3. currently the server will stop caching if the disk which cache_dir on usage reach 70%.

### CODE REVIEW

**Comment by Bodong-Yang** [comment]
This server requires `uvicorn` package to run. The following minimum script can be used to run the server.
```python
from server.server_app import App
import logging

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)
# enable_https means mainECU will use https to fetch files from the internet
app = App(cache_enabled=True, enable_https=False)

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="10.6.65.3", port=8080, log_level="info", lifespan="on")
```

**Comment by Bodong-Yang** [comment]
The usage of this proxy server is as follow:
```python
# use requests as example
import requests

url = ""
proxies = {
    'http': 'http://<proxy_ip>:<port>',
    'https': '',
}
response = requests.get(url, proxies=proxies)
```

**Comment by obi-t4** [comment]
I faced this error but how to reproduce it is unknown.
```
~/repos/ota-client.yang ((852d10b...))$ python server.py 
DEBUG:asyncio:Using selector: EpollSelector
INFO:     Started server process [36981]
INFO:uvicorn.error:Started server process [36981]
INFO:     Waiting for application startup.
INFO:uvicorn.error:Waiting for application startup.
INFO:ota_proxy.server_app:start ota http proxy app...
DEBUG:ota_proxy.ota_cache:init ota cache(cache_enabled=True, init=True, upper_proxy=None)
DEBUG:ota_proxy.db:init database...
DEBUG:ota_proxy.db:init sqlite database...
INFO:     Application startup complete.
INFO:uvicorn.error:Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:uvicorn.error:Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
DEBUG:ota_proxy.server_app:receive request for url='http://localhost:8081/dc.py'
DEBUG:ota_proxy.ota_cache:try to retrieve file from url='http://localhost:8081/dc.py'
DEBUG:ota_proxy.ota_cache:try to download and cache url='http://localhost:8081/dc.py'
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8081
DEBUG:urllib3.connectionpool:http://localhost:8081 "GET /dc.py HTTP/1.1" 200 1109
DEBUG:ota_proxy.ota_cache:new OTAFile request: http://localhost:8081/dc.py
DEBUG:ota_proxy.ota_cache:start to cache for http://localhost:8081/dc.py...
INFO:     127.0.0.1:39068 - "GET http%3A//localhost%3A8081/dc.py HTTP/1.1" 200 OK
DEBUG:ota_proxy.ota_cache:successfully cache http://localhost:8081/dc.py
DEBUG:ota_proxy.ota_cache:commit cache for http://localhost:8081/dc.py...


DEBUG:ota_proxy.server_app:receive request for url='http://localhost:8081/dc.py'
DEBUG:ota_proxy.ota_cache:try to retrieve file from url='http://localhost:8081/dc.py'
DEBUG:ota_proxy.ota_cache:cache hit for url='http://localhost:8081/dc.py'
, cache_meta=CacheMeta(url='http://localhost:8081/dc.py', hash='85aa765fe7a3dc9b', size=1109.0, content_type='text/plain', content_encoding='')
ERROR:ota_proxy.ota_cache:dangling cache entry found: 
cache_meta=CacheMeta(url='http://localhost:8081/dc.py', hash='85aa765fe7a3dc9b', size=1109.0, content_type='text/plain', content_encoding='')
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 377, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 145, in __call__
    await self.app(scope, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 126, in app
    await self._pull_data_and_send(url, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 74, in _pull_data_and_send
    f: ota_cache.OTAFile = await self._ota_cache.retrieve_file(url)
  File "/home/obi/repos/ota-client.yang/ota_proxy/ota_cache.py", line 416, in retrieve_file
    self._db.remove_urls(url)
  File "/home/obi/repos/ota-client.yang/ota_proxy/db.py", line 105, in remove_urls
    cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE url=?", urls)
sqlite3.ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 27 supplied.
ERROR:uvicorn.error:Exception in ASGI application
Traceback (most recent call last):
  File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 377, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 145, in __call__
    await self.app(scope, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 126, in app
    await self._pull_data_and_send(url, send)
  File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 74, in _pull_data_and_send
    f: ota_cache.OTAFile = await self._ota_cache.retrieve_file(url)
  File "/home/obi/repos/ota-client.yang/ota_proxy/ota_cache.py", line 416, in retrieve_file
    self._db.remove_urls(url)
  File "/home/obi/repos/ota-client.yang/ota_proxy/db.py", line 105, in remove_urls
    cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE url=?", urls)
sqlite3.ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 27 supplied.
INFO:     127.0.0.1:39070 - "GET http%3A//localhost%3A8081/dc.py HTTP/1.1" 500 Internal Server Error
^CINFO:     Shutting down
INFO:uvicorn.error:Shutting down
INFO:     Waiting for application shutdown.
INFO:uvicorn.error:Waiting for application shutdown.
INFO:ota_proxy.server_app:stopping ota http proxy app...
DEBUG:ota_proxy.ota_cache:shutdown ota-cache(cleanup=False)...
DEBUG:ota_proxy.db:closing db...
INFO:ota_proxy.ota_cache:shutdown ota-cache completed
INFO:ota_proxy.server_app:shutdown server completed
INFO:     Application shutdown complete.
INFO:uvicorn.error:Application shutdown complete.
INFO:     Finished server process [36981]
INFO:uvicorn.error:Finished server process [36981]
~/repos/ota-client.yang ((852d10b...))$ rm /tmp/ota-cache/cache_db 
~/repos/ota-client.yang ((852d10b...))$ python server.py 
DEBUG:asyncio:Using selector: EpollSelector
INFO:     Started server process [37126]
INFO:uvicorn.error:Started server process [37126]
INFO:     Waiting for application startup.
INFO:uvicorn.error:Waiting for application startup.
INFO:ota_proxy.server_app:start ota http proxy app...
DEBUG:ota_proxy.ota_cache:init ota cache(cache_enabled=True, init=True, upper_proxy=None)
DEBUG:ota_proxy.db:init database...
DEBUG:ota_proxy.db:init sqlite database...
INFO:     Application startup complete.
INFO:uvicorn.error:Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:uvicorn.error:Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:uvicorn.error:Shutting down
INFO:     Waiting for application shutdown.
INFO:uvicorn.error:Waiting for application shutdown.
INFO:ota_proxy.server_app:stopping ota http proxy app...
DEBUG:ota_proxy.ota_cache:shutdown ota-cache(cleanup=False)...
DEBUG:ota_proxy.db:closing db...
INFO:ota_proxy.ota_cache:shutdown ota-cache completed
INFO:ota_proxy.server_app:shutdown server completed
INFO:     Application shutdown complete.
INFO:uvicorn.error:Application shutdown complete.
INFO:     Finished server process [37126]
INFO:uvicorn.error:Finished server process [37126]
```

File: ota_proxy/ota_cache.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+                    logger.error(f"timeout caching for {self.meta.url}, abort")
+                    break
+
+                self._hash_f.update(data)
+                self.meta.size += self._dst_fp.write(data)
+
+            self._dst_fp.close()
+            if self.finished and self.meta.size > 0:  # not caching 0 size file
+                # rename the file to the hash value
+                hash = self._hash_f.hexdigest()[:16]
```
Comment (obi-t4) [inline ota_proxy/ota_cache.py:153]: whole hash value should be used. sha256 is unique in 256 bits.
```suggestion
                hash = self._hash_f.hexdigest()
```

**Review by obi-t4 — COMMENTED** [review]
@Bodong-Yang 
I left some minor comments. But I'm yet in review.

**Comment by Bodong-Yang** [comment]
> I faced this error but how to reproduce it is unknown.
> 
> ```
> ~/repos/ota-client.yang ((852d10b...))$ python server.py 
> DEBUG:asyncio:Using selector: EpollSelector
> INFO:     Started server process [36981]
> INFO:uvicorn.error:Started server process [36981]
> INFO:     Waiting for application startup.
> INFO:uvicorn.error:Waiting for application startup.
> INFO:ota_proxy.server_app:start ota http proxy app...
> DEBUG:ota_proxy.ota_cache:init ota cache(cache_enabled=True, init=True, upper_proxy=None)
> DEBUG:ota_proxy.db:init database...
> DEBUG:ota_proxy.db:init sqlite database...
> INFO:     Application startup complete.
> INFO:uvicorn.error:Application startup complete.
> INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
> INFO:uvicorn.error:Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
> DEBUG:ota_proxy.server_app:receive request for url='http://localhost:8081/dc.py'
> DEBUG:ota_proxy.ota_cache:try to retrieve file from url='http://localhost:8081/dc.py'
> DEBUG:ota_proxy.ota_cache:try to download and cache url='http://localhost:8081/dc.py'
> DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8081
> DEBUG:urllib3.connectionpool:http://localhost:8081 "GET /dc.py HTTP/1.1" 200 1109
> DEBUG:ota_proxy.ota_cache:new OTAFile request: http://localhost:8081/dc.py
> DEBUG:ota_proxy.ota_cache:start to cache for http://localhost:8081/dc.py...
> INFO:     127.0.0.1:39068 - "GET http%3A//localhost%3A8081/dc.py HTTP/1.1" 200 OK
> DEBUG:ota_proxy.ota_cache:successfully cache http://localhost:8081/dc.py
> DEBUG:ota_proxy.ota_cache:commit cache for http://localhost:8081/dc.py...
> 
> 
> DEBUG:ota_proxy.server_app:receive request for url='http://localhost:8081/dc.py'
> DEBUG:ota_proxy.ota_cache:try to retrieve file from url='http://localhost:8081/dc.py'
> DEBUG:ota_proxy.ota_cache:cache hit for url='http://localhost:8081/dc.py'
> , cache_meta=CacheMeta(url='http://localhost:8081/dc.py', hash='85aa765fe7a3dc9b', size=1109.0, content_type='text/plain', content_encoding='')
> ERROR:ota_proxy.ota_cache:dangling cache entry found: 
> cache_meta=CacheMeta(url='http://localhost:8081/dc.py', hash='85aa765fe7a3dc9b', size=1109.0, content_type='text/plain', content_encoding='')
> ERROR:    Exception in ASGI application
> Traceback (most recent call last):
>   File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 377, in run_asgi
>     result = await app(self.scope, self.receive, self.send)
>   File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
>     return await self.app(scope, receive, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 145, in __call__
>     await self.app(scope, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 126, in app
>     await self._pull_data_and_send(url, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 74, in _pull_data_and_send
>     f: ota_cache.OTAFile = await self._ota_cache.retrieve_file(url)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/ota_cache.py", line 416, in retrieve_file
>     self._db.remove_urls(url)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/db.py", line 105, in remove_urls
>     cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE url=?", urls)
> sqlite3.ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 27 supplied.
> ERROR:uvicorn.error:Exception in ASGI application
> Traceback (most recent call last):
>   File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 377, in run_asgi
>     result = await app(self.scope, self.receive, self.send)
>   File "/home/obi/.local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
>     return await self.app(scope, receive, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 145, in __call__
>     await self.app(scope, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 126, in app
>     await self._pull_data_and_send(url, send)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/server_app.py", line 74, in _pull_data_and_send
>     f: ota_cache.OTAFile = await self._ota_cache.retrieve_file(url)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/ota_cache.py", line 416, in retrieve_file
>     self._db.remove_urls(url)
>   File "/home/obi/repos/ota-client.yang/ota_proxy/db.py", line 105, in remove_urls
>     cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE url=?", urls)
> sqlite3.ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 27 supplied.
> INFO:     127.0.0.1:39070 - "GET http%3A//localhost%3A8081/dc.py HTTP/1.1" 500 Internal Server Error
> ^CINFO:     Shutting down
> INFO:uvicorn.error:Shutting down
> INFO:     Waiting for application shutdown.
> INFO:uvicorn.error:Waiting for application shutdown.
> INFO:ota_proxy.server_app:stopping ota http proxy app...
> DEBUG:ota_proxy.ota_cache:shutdown ota-cache(cleanup=False)...
> DEBUG:ota_proxy.db:closing db...
> INFO:ota_proxy.ota_cache:shutdown ota-cache completed
> INFO:ota_proxy.server_app:shutdown server completed
> INFO:     Application shutdown complete.
> INFO:uvicorn.error:Application shutdown complete.
> INFO:     Finished server process [36981]
> INFO:uvicorn.error:Finished server process [36981]
> ~/repos/ota-client.yang ((852d10b...))$ rm /tmp/ota-cache/cache_db 
> ~/repos/ota-client.yang ((852d10b...))$ python server.py 
> DEBUG:asyncio:Using selector: EpollSelector
> INFO:     Started server process [37126]
> INFO:uvicorn.error:Started server process [37126]
> INFO:     Waiting for application startup.
> INFO:uvicorn.error:Waiting for application startup.
> INFO:ota_proxy.server_app:start ota http proxy app...
> DEBUG:ota_proxy.ota_cache:init ota cache(cache_enabled=True, init=True, upper_proxy=None)
> DEBUG:ota_proxy.db:init database...
> DEBUG:ota_proxy.db:init sqlite database...
> INFO:     Application startup complete.
> INFO:uvicorn.error:Application startup complete.
> INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
> INFO:uvicorn.error:Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
> ^CINFO:     Shutting down
> INFO:uvicorn.error:Shutting down
> INFO:     Waiting for application shutdown.
> INFO:uvicorn.error:Waiting for application shutdown.
> INFO:ota_proxy.server_app:stopping ota http proxy app...
> DEBUG:ota_proxy.ota_cache:shutdown ota-cache(cleanup=False)...
> DEBUG:ota_proxy.db:closing db...
> INFO:ota_proxy.ota_cache:shutdown ota-cache completed
> INFO:ota_proxy.server_app:shutdown server completed
> INFO:     Application shutdown complete.
> INFO:uvicorn.error:Application shutdown complete.
> INFO:     Finished server process [37126]
> INFO:uvicorn.error:Finished server process [37126]
> ```

OMG I forgot to wrap each url in a tuple.
`urls = [(u, ) for u in urls]` should fix the problem.

File: ota_proxy/ota_cache.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        except KeyError:
+            return
+
+    def reserve_space(self, size: int) -> list:
+        enough_space = False
+        with self._lock:
+            hash_list, files_list = [], []
+            space_available = None
+            for h in self:
+                if space_available >= size:
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply ota_proxy/ota_cache.py:70]: This is a mistake, I will fix it.

File: ota_proxy/ota_cache.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+                try:
+                    data = self._queue.get(timeout=360)
+                except Exception:
+                    logger.error(f"timeout caching for {self.meta.url}, abort")
+                    break
+
+                self._hash_f.update(data)
+                self.meta.size += self._dst_fp.write(data)
+
+            self._dst_fp.close()
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply ota_proxy/ota_cache.py:150]: Below this line I rename the tmp cache entry to its sha256 value, I am afraid that renaming a file without closing the file descriptor is a bad idea.
Multiple times of closing file descriptor is safe as call close() on already closed file does nothing.

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

**Pull request merged**